src/bitmap/backend/atomic_bitmap_mmap.rs

```diff
@@ -0,0 +1,608 @@
+        // the region's starting address is not guarantee to be LOG_BLOCK_SIZE-page aligned
+        // which makes the implementation needlessly cumbersome.
+        // Note: The specification does not define whether the offset must be page-aligned or not.
+        // But, since we are receiving the offset from the frontend to be used to call mmap,
+        // we assume it is properly aligned (currently, qemu always send a 0 offset).
+        let logmem = MmapLogReg::from_file(fd, offset, len)?;
+
+        // We need to check that `log.as_ptr().offset(len)` doesn't wrap around and is
+        // contained within the mapped memory region.
+        check_pointer_offset(logmem.base_ptr(), len)?;
+
+        // The frontend sends a single bitmap (i.e., the log memory to be mapped using `fd`,
+        // `mmap_offset` and `mmap_size`) that covers the entire guest memory.
+        // However, since each memory region requires a bitmap relative to them, we have to
+        // adjust the offset and size, in number of pages, of this region.
+        let offset_pages = page_number(region_start_addr);
+        let size_page = page_number(region_len);
+
+        Ok(Self {
+            logmem,
+            pages_before_region: offset_pages,
+            number_of_pages: size_page,
+        })
+    }
+}
+
+// Just a handy object to manage the RAII of the mmap region.
+#[derive(Debug)]
+struct MmapLogReg {
+    addr: *const AtomicU8,
+    len: usize,
+}
+
+// SAFETY: Send is not automatically implemented because the raw pointer.
+// No one besides `MmapLogReg` has the raw pointer, so we can safely transfer it to another thread.
+unsafe impl Send for MmapLogReg {}
+
+// SAFETY: Sync is not automatically implemented because the raw pointer.
+// `MmapLogReg` doesn't have any interior mutability and all access to `&AtomicU8`
+// are done through atomic operations.
+unsafe impl Sync for MmapLogReg {}
+
+impl MmapLogReg {
+    fn from_file(fd: BorrowedFd, offset: isize, len: usize) -> io::Result<Self> {
+        #[cfg(not(miri))]
+        // SAFETY: `fd` is a valid file descriptor and we are not using `libc::MAP_FIXED`.
+        let addr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                len as libc::size_t,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+
+        #[cfg(not(miri))]
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        #[cfg(miri)]
+        if len == 0 {
+            return Err(io::Error::from_raw_os_error(libc::EINVAL));
+        }
+
+        // Miri does not support the mmap syscall, so we use rust's allocator for miri tests
+        #[cfg(miri)]
+        let addr = unsafe {
+            std::alloc::alloc_zeroed(std::alloc::Layout::from_size_align(len, 1).unwrap())
+        };
```

Comment:
This `#[cfg]` use makes it non-trivial to see that we aren’t potentially leaking the memory region returned by `mmap()`. In general, I find it confusing to mix code for two targets that does completely different things.

Because these are two fully distinct implementations, I’d prefer making that explicit, i.e.
```rust
#[cfg(not(miri))]
fn from_file(fd: BorrowedFd, offset: isize, len: usize) -> io::Result<Self> { /* mmap */ }
#[cfg(miri)]
fn from_file(fd: BorrowedFd, offset: isize, len: usize) -> io::Result<Self> { /* alloc_zeroed */ }
```

Or at least
```rust
let addr = if cfg!(not(miri)) { ... } else { ... };
```

Reply:
Ok